File: backend/app/models/deck.py
# SPDX-License-Identifier: AGPL-3.0-or-later
import uuid

from sqlalchemy import ForeignKey, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.mixins import TimestampedModel


class Deck(TimestampedModel):
    """A user's custom deck, or a system deck (e.g. Rider-Waite-Smith) when `user_id` is null."""

    __tablename__ = "decks"

    name: Mapped[str] = mapped_column(Text)
    description: Mapped[str | None] = mapped_column(Text)
    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
    )

File: frontend/packages/api-client/src/models/deck.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
import { Page } from "./pagination";

export type DeckType = "system" | "custom";

export interface Deck {
  id: string;
  name: string;
  description: string | null;
  user_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface AdminDeck extends Deck {
  owner_username: string | null;
}

export type PaginatedDecks = Page<AdminDeck>;
